### ISSUE TITLE
seg fault when using Faiss with multi-GPU after torch imported

### ISSUE THREAD

**Issue by senhuiguo** [facebookresearch/faiss#793]
package version: python=1.0, faiss-gpu=1.5
I get segmentation fault when using GPU index after torch is imported

```python
import torch
import faiss
res_bag = [faiss.StandardGpuResources() for idx in range(2)] # num_gpu >= 2
cpu_index = faiss.IndexFlatIP(256)
co = faiss.GpuMultipleClonerOptions()
co.useFloat16 = True
gpu_index = faiss.index_cpu_to_gpu_multiple_py(res_bag, cpu_index, co)
```

however, if I imported faiss first, the error was gone
```python
import faiss
import torch
res_bag = [faiss.StandardGpuResources() for idx in range(2)]
cpu_index = faiss.IndexFlatIP(256)
co = faiss.GpuMultipleClonerOptions()
co.useFloat16 = True
gpu_index = faiss.index_cpu_to_gpu_multiple_py(res_bag, cpu_index, co)
```

I was wondering if there is a way to use faiss multi-gpu index after torch imported

**Comment by beauby** [comment]
This is a known issue that is fixed with the latest versions of Faiss (v1.5.1) and pytorch.

**Comment by senhuiguo** [comment]
@beauby Are you sure?
```bash
faiss-gpu                 1.5.1            py37h32c130d_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
pytorch                   1.0.1           py3.7_cuda9.0.176_cudnn7.4.2_2    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
torchvision               0.2.2                      py_3    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
```

**Comment by beauby** [comment]
@senhuiguo Could you post your entire conda env (`conda list`)?

**Comment by senhuiguo** [comment]
@beauby Here you go:
```bash
# packages in environment at /mnt/lustre/guosenhui/anaconda3/envs/mixed3.7:
#
# Name                    Version                   Build  Channel
backcall                  0.1.0                      py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
blas                      1.0                         mkl  
bzip2                     1.0.6             h14c3975_1002    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
ca-certificates           2019.3.9             hecc5488_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
certifi                   2019.3.9                 py37_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
cffi                      1.12.2           py37hf0e25f4_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
cudatoolkit               9.0                  h13b8566_0  
cycler                    0.10.0                     py_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
dbus                      1.13.2               h714fa37_1  
decorator                 4.4.0                      py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
expat                     2.2.5             hf484d3e_1002    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
faiss-gpu                 1.5.1            py37h32c130d_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
fontconfig                2.13.1            he4413a7_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
freetype                  2.10.0               he983fc9_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
gettext                   0.19.8.1          hc5be6a0_1002    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
glib                      2.56.2            had28632_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
gst-plugins-base          1.14.0               hbbd80ab_1  
gstreamer                 1.14.0               hb453b48_1  
icu                       58.2              hf484d3e_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
intel-openmp              2019.3                      199  
ipython                   7.4.0            py37h24bf2e0_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
ipython_genutils          0.2.0                      py_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
isodate                   0.6.0                    pypi_0    pypi
jedi                      0.13.3                   py37_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
jpeg                      9c                h14c3975_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
kiwisolver                1.0.1           py37h6bb024c_1002    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libblas                   3.8.0           5_h6e990d7_netlib    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libcblas                  3.8.0           5_h6e990d7_netlib    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libffi                    3.2.1             he1b5a44_1006    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libgcc-ng                 8.2.0                hdf63c60_1  
libgfortran-ng            7.3.0                hdf63c60_0  
libiconv                  1.15              h516909a_1005    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
liblapack                 3.8.0           5_h6e990d7_netlib    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libpng                    1.6.36            h84994c4_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libprotobuf               3.7.1                h8b12597_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libstdcxx-ng              8.2.0                hdf63c60_1  
libtiff                   4.0.10            h648cc4a_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libuuid                   2.32.1            h14c3975_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libxcb                    1.13              h14c3975_1002    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
libxml2                   2.9.8             h143f9aa_1005    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
matplotlib                3.0.3                    py37_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
matplotlib-base           3.0.3            py37h167e16e_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
mkl                       2019.3                      199  
mkl_fft                   1.0.11           py37h14c3975_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
mkl_random                1.0.2            py37h637b7d7_2    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
ncurses                   6.1               hf484d3e_1002    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
networkx                  2.3                      pypi_0    pypi
ninja                     1.9.0                h6bb024c_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
numpy                     1.16.2           py37h7e9f1db_0  
numpy-base                1.16.2           py37hde5b4d6_0  
olefile                   0.46                       py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
openssl                   1.1.1b               h14c3975_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pandas                    0.24.2                   pypi_0    pypi
parso                     0.4.0                      py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pcre                      8.43                 he6710b0_0  
pexpect                   4.7.0                    py37_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pickleshare               0.7.5                 py37_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pillow                    6.0.0            py37he7afcd5_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pip                       19.0.3                   py37_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
plyfile                   0.7                      pypi_0    pypi
prompt_toolkit            2.0.9                      py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
protobuf                  3.7.1            py37he1b5a44_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
psutil                    5.6.1            py37h14c3975_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pthread-stubs             0.4               h14c3975_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
ptyprocess                0.6.0                   py_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pycparser                 2.19                     py37_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pygments                  2.3.1                      py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pyparsing                 2.4.0                      py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pyqt                      5.6.0           py37h13b7fb3_1008    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
python                    3.7.3                h5b0a415_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
python-dateutil           2.8.0                      py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
pytorch                   1.0.1           py3.7_cuda9.0.176_cudnn7.4.2_2    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
pytz                      2019.1                   pypi_0    pypi
qt                        5.6.3                h8bf5577_3  
rdflib                    4.2.2                    pypi_0    pypi
readline                  7.0               hf8c457e_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
scikit-learn              0.20.3                   pypi_0    pypi
scipy                     1.2.1                    pypi_0    pypi
setuptools                41.0.0                   py37_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
sip                       4.18.1          py37hf484d3e_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
six                       1.12.0                py37_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
sqlite                    3.26.0            h67949de_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
tk                        8.6.9             h84994c4_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
torch-cluster             1.2.4                    pypi_0    pypi
torch-geometric           1.1.2                    pypi_0    pypi
torch-scatter             1.1.2                    pypi_0    pypi
torch-sparse              0.2.4                    pypi_0    pypi
torch-spline-conv         1.0.6                    pypi_0    pypi
torchvision               0.2.2                      py_3    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/pytorch
tornado                   6.0.2            py37h516909a_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
tqdm                      4.31.1                     py_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
traitlets                 4.3.2                 py37_1000    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
wcwidth                   0.1.7                      py_1    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
wheel                     0.33.1                   py37_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
xorg-libxau               1.0.9                h14c3975_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
xorg-libxdmcp             1.1.3                h516909a_0    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
xz                        5.2.4             h14c3975_1001    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
zlib                      1.2.11            h14c3975_1004    https://mirrors.tuna.tsinghua.edu.cn/anaconda/cloud/conda-forge
```

**Comment by beauby** [comment]
After investigation, the problem is only partially fixed: it is still failing if cuda is installed and in the loader path (in which case pytorch loads the system cuda instead of the one shipped with the `cudatoolkit` package).

**Comment by beauby** [comment]
Update: I believe this is related to https://github.com/pytorch/pytorch/issues/19309.

**Comment by beauby** [comment]
Closing, as this is fixed (tested with faiss-gpu 1.5.2 and pytorch-nightly 1.2.0.dev20190530). Feel free to reopen if needed.
